### ISSUE TITLE
Unhandled rejection on rollback

### ISSUE THREAD

**Issue by magorn** [tgriesser/knex#3262]
[log.txt](https://github.com/tgriesser/knex/files/3261698/log.txt)

# Environment
 "knex": "^0.17.0"
"pg": "^7.11.0",
Database + version: Postgres 11.3
OS: Mac OS X

# Bug

1. Expects a clean roll back

2. Error message
```
Unhandled rejection error: column "acccountnumber" of relation "account" does not exist
    at Connection.parseE (/Users/magnusg/dev/nodejs/crudPractice-api/node_modules/pg/lib/connection.js:602:11)
    at Connection.parseMessage (/Users/magnusg/dev/nodejs/crudPractice-api/node_modules/pg/lib/connection.js:399:19)
    at Socket.<anonymous> (/Users/magnusg/dev/nodejs/crudPractice-api/node_modules/pg/lib/connection.js:121:22)
    at Socket.emit (events.js:189:13)
    at addChunk (_stream_readable.js:284:12)
    at readableAddChunk (_stream_readable.js:265:11)
    at Socket.Readable.push (_stream_readable.js:220:10)
    at TCP.onStreamRead [as onread] (internal/stream_base_commons.js:94:17)
```

3. 
#run against database
```
create table customer
(
  customernumber varchar(20) not null
    constraint pk_customer
      primary key,
  firstname      varchar(35),
  lastname       varchar(35)
);

create table account
(
  accountnumber  numeric(8) not null
    constraint pk_accountnumber
      primary key,
  customernumber varchar(20)
);
```

 #Code
```
var database = require('knex')({
    client: 'pg',
    connection: {
        host: '127.0.0.1',
        user: 'nodejs',
        password: 'nodejs',
        database: 'crud1'
    },
    syncStackTraces: true
});

const createAccount = async () => {
    const trxProvider = database.transactionProvider();
    const trx = await trxProvider();

    try {
        var customerData = {
            customernumber: '196210010102',
            firstname: "magnus",
            lastname: "andersson"
        };
        await trx('customer').insert(customerData).catch(error1 => {
            throw error1
        });
        var accounData = {acccountnumber: 202, customernumber: customerData.customernumber};
        await trx('account').insert(accounData).catch(error1 => {
            throw error1
        });
        await trx.commit();
        return Promise.resolve(customerData);

    } catch (err) {
        console.log('rolling back');
        await trx.rollback(err);
        return Promise.reject(err);

    }
}

createAccount().then(
    result => console.log(result)).catch(error => {
    console.log(error);
});

```

**Comment by elhigu** [comment]
`column "acccountnumber" of relation "account" does not exist`... comes from database, meaning that your account table doesn't have column named accountnumber.

Please add a row to your account table in setup code and before doing anything else run 

```
console.log(await knex('account'));
```

I suspect your DB is not what you think it is and that will rule it out if it works.

also you could checkout what https://knexjs.org/#Builder-columnInfo
`await knex('account').columnInfo()`returns.

Lets reopen if this seems to be bug in knex.

ps. you don't have to wrap your return values to Promise.resolve / Promise.reject in async functions. You can just return anything / throw an error and it will be implicitly wrapped to promise.

**Comment by magorn** [comment]
I know that the column is faulty that is the test case, the code works if I change acccountnumber to acountnumber
 var accounData = {acccountnumber: 202, customernumber: customerData.customernumber};
I expect the rollback to rollback everything cleanly without , which is the customer insert, so I get all or nothing.

**Comment by elhigu** [comment]
Right, ok so this is duplicate of https://github.com/tgriesser/knex/issues/3246

**Comment by magorn** [comment]
No I read that one and I do think I have handler for both outcomes, how do I code to not get the error when I am doing rollback?
createAccount().then(
    result => console.log(result)).catch(error => {
    console.log(error);
});

**Comment by elhigu** [comment]
afaik it is the same problem and currently there is no way to fix it if I understood the last comment from @kibertoad right. So there is an internal promise left which rejects/resolves on rollback / commit, which doesn't have handler for errors.

I suppose for now you need to use transactions like this:

```

const createAccount = async () => {
    try {
        // in this case await is actually necessary to be able to write the outer try / catch
        return await database.transaction(async trx => {
            var customerData = {
                customernumber: '196210010102',
                firstname: "magnus",
                lastname: "andersson"
            };
            await trx('customer').insert(customerData);
            var accounData = {acccountnumber: 202, customernumber: customerData.customernumber};
            await trx('account').insert(accounData);
            return customerData;
        });
    } catch (e) {
        // added try / catch just to underline the fact that transaction is automatically rolled back
        // if exceptions were thrown from transaction handler
        console.log('Transaction was rolled back');
        throw e;
    }
}
```

**Comment by magorn** [comment]
Great, Thank You for the rewrite and explanations,  one other thing why can I not get the real stack ,
I mean rownumber with the failing row below  and call stack?
var accounData = {acccountnumber: 202, customernumber: customerData.customernumber};

**Comment by magorn** [comment]
hmm, guess,  that parseE
msg = new Error(fields.M)
is the real error, which then is the subexception

**Comment by elhigu** [comment]
Try to set `asyncStackTraces: true` to config, now there seems to be a typo. It should record stack trace where query builder was created and return that when DB errors are thrown.

**Comment by magorn** [comment]
Yes, that made the trick,Thanks.

**Comment by magorn** [comment]
Does knex handle sequencers? I tried below with no luck 
 var nextAccountNumber = await trx.raw('SELECT nextval(\'account_sequencer)\'');
#DB setup
CREATE SEQUENCE account_sequencer START 200100;
SELECT nextval('account_sequencer');

**Comment by elhigu** [comment]
you have invalid quotes there

**Comment by magorn** [comment]
Yes, thanks, works!

**Comment by kibertoad** [comment]
@magorn Transaction rollback handling was improved in master. Would releasing 0.18.0-next3 with this change be helpful for you?

**Comment by magorn** [comment]
Yes, great!

**Comment by kibertoad** [comment]
Resolved in 0.18.0
